File: scala/ch01/ex1.46.scala
val tolerance = 0.00001

def square(x: Double) = x*x

def iterativeImprove(goodEnough: (Double) => Boolean, improve: (Double) => Double) = {
  def iterate(guess2: Double): Double = goodEnough(guess2) match {
    case true => guess2
    case _ => iterate(improve(guess2))
  }
  (guess:Double) => iterate(guess)
}

def average(x: Double, y: Double) = (x + y) / 2.0

def goodEnough(guess: Double, x: Double) = Math.abs(square(guess) - x) < tolerance

def improve(guess: Double, x: Double) = average(guess, x / guess)
  
def sqrt(x: Double) = iterativeImprove(
  (guess:Double) => goodEnough(guess, x), 
  (guess:Double) => improve(guess, x))(1.0)
    
import org.scalatest._
import org.scalatest.matchers._

object iterativeImproveForSqrtSpec extends Spec with ShouldMatchers {
  describe ("nth-power calculator") {
    it ("should calculate the correct values") {
      sqrt(2)    should be (1.414213562373095 plusOrMinus 0.0001)
      sqrt(3)    should be (1.732050807568877 plusOrMinus 0.0001)
      sqrt(4)    should be (2.0               plusOrMinus 0.0001)
      sqrt(4E16) should be (2E8               plusOrMinus 1.0)
    }
  }
}
iterativeImproveForSqrtSpec execute

File: clojure/ch01/ex1.46.clj
(ns sicp.ch01 (:use clojure.test))

(defn iterative-improve [good-enough? improve]
  (fn [guess] 
    (loop [guess2 guess]
      (if (good-enough? guess2)
        guess2
        (recur (improve guess2))))))

(defn square [x] (* x x))

(defn average [x y]
  (/ (+ x y) 2))

(defn good-enough? [guess x]
  (< (Math/abs (- (square guess) x)) 0.001))

(defn improve [guess x]
  (average guess (/ x guess)))
  
(defn sqrt [x]
  ((iterative-improve
    #(good-enough? % x)
    #(improve % x)) 1.0))
    
; Is there a method like this in test-is?? Couldn't find something like it documented.
(defn is-within [expected actual delta]
  (< (Math/abs (- expected actual)) delta))

(deftest test-iterative-improve-for-sqrt
  (is (is-within 1.414213562373095 (sqrt 2)    0.0001))
  (is (is-within 1.732050807568877 (sqrt 3)    0.0001))
  (is (is-within 2.0               (sqrt 4)    0.0001))
  (is (is-within 2E8               (sqrt 4E16) 1.0))
  ; For very small X, the original sqrt is horrible; this test fails
  ; (check-= (sqrt 4E-16) 2E-8 1E-12)
  )
(run-tests)